feat: add support for encode&decode

BanBuDu0 opened the pull request:
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?


According to https://github.com/risingwavelabs/risingwave/issues/8937.
The encode and decode function are implemented in this pull request.

<!--

**This section will be used as the commit message. Please do not leave this empty!**

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- Summarize your change (**mandatory**)
- How does this PR work? Need a brief introduction for the changed logic (optional)
- Describe clearly one logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)
- Refer to a related PR or issue link (optional)


-->

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [x] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.
<!-- To manually trigger a benchmark, please check out [Notion](https://www.notion.so/risingwave-labs/Manually-trigger-nexmark-performance-dashboard-test-b784f1eae1cf48889b2645d020b6b7d3). -->

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<!-- 

You can ignore or delete the section below if you ticked the checkbox above.

Otherwise, remove the checkbox above and write a release note below.

-->

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

<!--
Please create a release note for your changes. 

Discuss technical details in the "What's changed" section, and 
focus on the impact on users in the release note.

You should also mention the environment or conditions where the impact may occur.
-->

</details>

lmatz commented:
Could you help enable the test cases here: https://github.com/risingwavelabs/risingwave/blob/0bdc988be5cb8b307e204815f9237c6949f664c9/src/tests/regress/data/sql/strings.sql#L649

BanBuDu0 commented:
> `SELECT encode(('\x' || repeat('1234567890abcdef0001', 7))::bytea, 'base64');` `SELECT decode(encode(('\x' || repeat('1234567890abcdef0001', 7))::bytea, 'base64'), 'base64');` These two sqls will produce an error: `ERROR: QueryError: Bind error: cannot cast type "varchar" to "bytea" in Explicit context`
> 
> Seems like an error happend in compile-time which cast expr bind error. I see that no Bytea type mentioned in `CAST_MAP` https://github.com/risingwavelabs/risingwave/blob/main/src/expr/src/sig/cast.rs#L51

I see that the cast function [varchar->bytea] was already implemented here https://github.com/risingwavelabs/risingwave/blob/main/src/expr/src/vector_op/cast.rs#L196

So I fix it by simple add Bytea type to CAST_MAP, see commit：https://github.com/risingwavelabs/risingwave/pull/9351/commits/432b7052926f65eebf2f17ef420ee03095dac9b4
Is there something wrong with this fix?

xiangjinwu commented on proto/expr.proto:
```diff
@@ -117,6 +117,8 @@ message ExprNode {
     SQRT = 256;
     DEGREES = 257;
     RADIANS = 258;
+    ENCODE = 259;
+    DECODE = 260;
```
Historically we have mixed both number-related and string-related expressions to 2xx. Could you put these 2 new string/bytea expressions starting 310 with a comment saying "string/bytea expressions"?

xiangjinwu commented on src/tests/regress/data/sql/strings.sql:
```diff
@@ -648,13 +648,12 @@ select md5('12345678901234567890123456789012345678901234567890123456789012345678
---@ SELECT encode('\x1234567890abcdef00', 'hex');
---@ SELECT decode('1234567890abcdef00', 'hex');
---@ SELECT encode(('\x' || repeat('1234567890abcdef0001', 7))::bytea, 'base64');
---@ SELECT decode(encode(('\x' || repeat('1234567890abcdef0001', 7))::bytea,
---@                      'base64'), 'base64');
```
Keep them on 2 lines with the leading indentation spaces. In other words, only remove the 4 leading chars of `--@ ` and avoid any other changes, or the test output would not match.

xiangjinwu reviewed (commented):
Thanks for your contribution. And your fix to the `CAST_MAP` is correct.

BanBuDu0 commented:
After I changed the CAST_MAP, I should also modify the e2e_test/batch/catalog/pg_cast.slt.part following the change of cast_map.
By the way, I also fixed the from_oid function here https://github.com/risingwavelabs/risingwave/blob/main/src/common/src/types/postgres_type.rs#L48, there was no Bytea before.
See https://github.com/risingwavelabs/risingwave/pull/9351/commits/186bc936037e940fae7fc79ebf2710c2b19022ce

Finally, all the test was passsed now🥳.

lmatz reviewed (approved):
Many thanks for your contribution!
